src/Engine.ts

```diff
@@ -709,3 +828,360 @@ export class Engine {
+
+    const validationFunc = async (graphNode: GraphNode): Promise<boolean> => {
+      if (validationMap.has(graphNode.graphID)) {
+        return validationMap.get(graphNode.graphID) || false
+      }
+
+      let parsedTx
+      try {
+        parsedTx = Transaction.fromHex(graphNode.rawTx)
+        if (graphNode.proof !== undefined) {
+          parsedTx.merklePath = MerklePath.fromHex(graphNode.proof)
+        }
+      } catch (error) {
+        console.error('Error parsing transaction or proof:', error)
+        return false
+      }
+
+      const validatedChildren: GraphNode[] = []
+      for (const child of graphNode.children) {
+        if (!validationMap.has(child.graphID)) {
+          const isValidChild = await validationFunc(child)
+          if (isValidChild) {
+            validatedChildren.push(child)
+          }
+        } else if (validationMap.get(child.graphID)) {
+          validatedChildren.push(child)
+        }
+      }
+
+      let admittanceResult
+      try {
+        admittanceResult = await this.engine.managers[this.topic].identifyAdmissibleOutputs(parsedTx.toBEEF(), validatedChildren.map(child => child.outputIndex))
+      } catch (error) {
+        console.error('Error in admittance check:', error)
+        return false
+      }
+
+      const isValid = admittanceResult.outputsToAdmit.includes(graphNode.outputIndex)
+      validationMap.set(graphNode.graphID, isValid)
+
+      if (isValid === false) {
+        return false
+      }
+
+      // Reached the root successfully
+      if (graphNode.parent === undefined) {
+        return true
+      }
+
+      return await validationFunc(graphNode.parent)
+    }
+
+    const tipNode = this.temporaryGraphNodeRefs[graphID]
+    if (tipNode === undefined) {
+      throw new Error(`Graph node with ID ${graphID} not found`)
+    }
+
+    const isValid = await validationFunc(tipNode)
+    if (!isValid) {
+      throw new Error('The graph is not well-anchored')
+    }
+  }
+
+  /**
+   * Deletes all data associated with a temporary graph that has failed to sync, if the graph exists.
+   * @param graphID The TXID and output index (in 36-byte format) for the UTXO at the tip of this graph.
+   */
+  async discardGraph(graphID: string): Promise<void> {
+    for (const [nodeId, graphRef] of Object.entries(this.temporaryGraphNodeRefs)) {
+      if (graphRef.graphID === graphID) {
+        // Delete child node
+        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+        delete this.temporaryGraphNodeRefs[nodeId]
+      }
+    }
+  }
+
+  /**
+   * Finalizes a graph, solidifying the new UTXO and its ancestors so that it will appear in the list of known UTXOs.
+   * @param graphID The TXID and output index (in 36-byte format) for the UTXO at the tip of this graph.
+   */
+  async finalizeGraph(graphID: string): Promise<void> {
+    // TODO: use similar function as validationFunc to construct the finalized graph with necessary components
```

Comment:
Assemble the BEEFs into the correct order, for submission through the topic manager. Send them all through similar logic to `Engine.submit()`, but without broadcasting to the network.

This will drive all lookup services, handle `previousCoins`, and store new outputs in the `Engine.storage` instance.

To achieve this, consider factoring out common logic from the inner loop of the `Engine.submit()` method, so that it can be used in both places.

Reply:
To create the BEEFs in the correct order, follow these steps:
- Create a `Set` with BEEFs in it.
- Recursively go back to the earliest transactions at each deepest point in the graph (do not use Promises / async logic for this)
- Once there, recurse back forward through, from each deepest point towards the parent.
- During the forward recursion, serialize a BEEF (we need a utility function that computes a BEEF given a `GraphNode`)
- After the BEEF is computed for the node, add it to the `Set` of BEEFs.
- Because this is a Set, no duplicative BEEFs will be added.
- Because this is not async, there will be no out-of-order BEEFs.
- After this process, convert the Set to an array (hopefully this preserves the order of the elements!)
- Then, after the final Array of ordered BEEFs is created, drive them through the topic.
- During the drive-through logic, be sure to respect `applied_transactions`! This will avoid multiple parallel syncs from running into each other on the same topic.